core/src/test/java/org/apache/calcite/test/RelMetadataTest.java

```diff
@@ -3412,4 +3414,41 @@ public String colType(RelNode rel, int column) {
     assertThat(columnOrigin.getOriginTable().getRowType().getFieldNames().get(5),
         equalTo("SAL"));
   }
+
+  // ----------------------------------------------------------------------
+  // Tests for SortCpuCost
+  // ----------------------------------------------------------------------
+
+  @Test void testSortCpuCostOffsetLimit() {
+    final String sql = "select ename from emp order by ename limit 5 offset 5";
+    double cpuCost = EMP_SIZE * Math.log(10) * 4;
+    checkCpuCost(sql, cpuCost);
+  }
+
+  @Test void testSortCpuCostLimit() {
+    final String sql = "select ename from emp limit 10";
+    checkCpuCost(sql, 0d);
+  }
+
+  @Test void testSortCpuCostLimit0() {
+    final String sql = "select ename from emp order by ename limit 0";
+    checkCpuCost(sql, 0d);
+  }
+
+  @Test void testSortCpuCostLargeLimit() {
+    final String sql = "select ename from emp order by ename limit 10000";
+    double cpuCost = EMP_SIZE * Math.log(EMP_SIZE) * 4;
+    checkCpuCost(sql, cpuCost);
+  }
+
+  private void checkCpuCost(String sql, double expected) {
+    RelNode rel = convertSql(sql);
+    assertThat("query should be sort", rel instanceof LogicalSort);
+    final RelMetadataQuery mq = rel.getCluster().getMetadataQuery();
+    RelOptPlanner planner = new VolcanoPlanner();
+    RelOptCost cost = rel.computeSelfCost(planner, mq);
+    assertThat(cost, notNullValue());
+    final double result = cost.getCpu();
+    assertThat(result, is(expected));
```

Comment:
Please add a clarification message that explains why a specific value expected (like above)

Reply:
Hi, I learned these notNullValue/is matcher from other test.  I found they will give a clarification message by themselves.  Take the notNullValue for example, when fail, the message like:
Expected: not null
     but: was null
java.lang.AssertionError: 
Expected: not null
     but: was null
I do not find a way to add a clarification message for these matcher

Reply:
Suppose I have never seen the test code, and I see the following failure in the CI output:

```
Expected: not null
but: was null
java.lang.AssertionError:
Expected: not null
but: was null
```

What should I do? Why "expected not null"? Why "was null"? What exactly was null?

The proper assertion failure should look like a good bug report rather than "go figure out what went wrong".

Bad assertions and bad test code make maintenance much harder.

Reply:
>I do not find a way to add a clarification message for these matcher

Please use `String reason` parameter: http://hamcrest.org/JavaHamcrest/javadoc/2.1/org/hamcrest/MatcherAssert.html#assertThat-java.lang.String-T-org.hamcrest.Matcher-

Reply:
Thanks, I recall that I can use assertEquals(double expected, double actual, Supplier<String> messageSupplier), just as
in previous pr you tell me.